src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/service/CaseAllocationService.kt

```diff
@@ -0,0 +1,110 @@
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.jpa.entity.CaseAllocationEntity
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.jpa.repository.CaseAllocationRepository
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.jpa.repository.PrisonerRepository
+import java.time.LocalDateTime
+
+@Service
+class CaseAllocationService(
+  private val prisonerRepository: PrisonerRepository,
+  private val caseAllocationRepository: CaseAllocationRepository,
+) {
+
+  @Transactional
+  fun getCaseAllocationByNomsId(
+    nomsId: String,
+  ): CaseAllocationEntity? {
+    val prisoner = prisonerRepository.findByNomsId(nomsId)
+      ?: throw ResourceNotFoundException("Prisoner with id $nomsId not found in database")
+    val caseAllocationEntity = caseAllocationRepository.findByPrisonerIdAndIsDeleted(
+      prisoner.id(),
+      false,
+    )
+    return caseAllocationEntity ?: throw ResourceNotFoundException("No officer assigned for prisoner with id $nomsId")
+  }
+
+  @Transactional
+  fun getAllCaseAllocationByStaffId(
+    staffId: Int,
+  ): List<CaseAllocationEntity?> {
+    val caseAllocationEntityList = caseAllocationRepository.findByStaffIdAndIsDeleted(
+      staffId,
+      false,
+    )
+    return caseAllocationEntityList
+  }
+
+  @Transactional
+  fun unAssignCase(caseAllocation: CaseAllocation): List<CaseAllocationEntity?> {
+    val caseList = emptyList<CaseAllocationEntity?>().toMutableList()
+    for (nomsId in caseAllocation.nomsIds) {
+      val prisoner = prisonerRepository.findByNomsId(nomsId)
+        ?: throw ResourceNotFoundException("Prisoner with id $nomsId not found in database")
+      val caseAllocationExists = caseAllocationRepository.findByPrisonerIdAndIsDeleted(
+        prisoner.id(), false,
+      ) ?: throw ResourceNotFoundException("Unable to unassign, no officer assigned for prisoner with id $nomsId")
+      val case = caseAllocationExists?.let { delete(it) }
+      caseList.add(case)
+    }
+    return caseList
+  }
+
+  @Transactional
+  fun delete(caseAllocation: CaseAllocationEntity): CaseAllocationEntity {
+    caseAllocation.isDeleted = true
+    caseAllocation.deletionDate = LocalDateTime.now()
+    caseAllocationRepository.save(caseAllocation)
+    return caseAllocation
+  }
+
+  @Transactional
+  fun assignCase(caseAllocation: CaseAllocation): MutableList<CaseAllocationEntity?> {
+    val caseList = emptyList<CaseAllocationEntity?>().toMutableList()
+    if (caseAllocation.staffId != null && caseAllocation.staffFirstName != null && caseAllocation.staffLastName != null) {
+      for (nomsId in caseAllocation.nomsIds) {
+        val prisoner = prisonerRepository.findByNomsId(nomsId)
+          ?: throw ResourceNotFoundException("Prisoner with id $nomsId not found in database")
+        val caseAllocationExists = caseAllocationRepository.findByPrisonerIdAndIsDeleted(
+          prisoner.id(),
+          false,
+        )
+        if (caseAllocationExists != null) {
```

Comment:
when we assign a case that's already assigned we'd expect it just to be assigned to the new person and the old record be deleted

Reply:
Done